nibabel/nicom/ascconv.py

```diff
@@ -0,0 +1,207 @@
+
+
+class Atom(object):
+    """ Object to hold operation, object type and object identifier
+
+    An atom represents an element in an expression.  For example::
+
+        a.b[0].c
+
+    has four elements.  We call these elements "atoms".
+
+    We represent objects (like ``a``) as dicts for convenience.
+
+    The last element (``.c``) is an ``op = ast.Attribute`` operation where the
+    object type (`obj_type`) of ``c`` is not constrained (we can't tell from
+    the operation what type it is).  The `obj_id` is the name of the object --
+    "c".
+
+    The second to last element ``[0]``, is ``op = ast.Subscript``, with object type
+    dict (we know from the subsequent operation ``.c`` that this must be an
+    object, we represent the object by a dict).  The `obj_id` is the index 0.
+
+    Parameters
+    ----------
+    op : {'name', 'attr', 'list'}
+        Assignment type.  Assignment to name (root namespace), attribute or
+        list element.
+    obj_type : {list, dict, other}
+        Object type being assigned to.
+    obj_id : str or int
+        Key (``obj_type is dict``) or index (``obj_type is list``)
+    """
+
+    def __init__(self, op, obj_type, obj_id):
+        self.op = op
+        self.obj_type = obj_type
+        self.obj_id = obj_id
+
+
+class NoValue(object):
+    """ Signals no value present """
+
+
+def assign2atoms(assign_ast, default_class=int):
+    """ Parse single assignment ast from ascconv line into atoms
+
+    Parameters
+    ----------
+    assign_ast : assignment statement ast
+        ast derived from single line of ascconv file.
+    default_class : class, optional
+        Class that will create an object where we cannot yet know the object
+        type in the assignment.
+
+    Returns
+    -------
+    atoms : list
+        List of :class:`atoms`.  See docstring for :class:`atoms`.  Defines
+        left to right sequence of assignment in `line_ast`.
+    """
+    if not len(assign_ast.targets) == 1:
+        raise AscconvParseError('Too many targets in assign')
+    target = assign_ast.targets[0]
+    atoms = []
+    prev_target_type = default_class  # Placeholder for any scalar value
+    while True:
+        if isinstance(target, ast.Name):
+            atoms.append(Atom(target, prev_target_type, target.id))
+            break
+        if isinstance(target, ast.Attribute):
+            atoms.append(Atom(target, prev_target_type, target.attr))
+            target = target.value
+            prev_target_type = OrderedDict
+        elif isinstance(target, ast.Subscript):
+            index = target.slice.value.n
+            atoms.append(Atom(target, prev_target_type, index))
+            target = target.value
+            prev_target_type = list
+        else:
+            raise AscconvParseError(
+                'Unexpected LHS element {0}'.format(target))
+    return reversed(atoms)
+
+
+def _create_obj_in(atom, root):
+    """ Create object defined in `atom` in dict-like given by `root`
+
```

Comment:
May be worth noting (I should have noted) that returns `atom` if there is already something with the same name in `root`.